java-checks/src/main/java/org/sonar/java/checks/NoTestInTestClassCheck.java

```diff
@@ -38,21 +39,46 @@
 import org.sonar.plugins.java.api.tree.Tree;
 import org.sonar.plugins.java.api.tree.Tree.Kind;
 
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
 import java.util.List;
-import java.util.function.Predicate;
+import java.util.Map;
 import java.util.stream.Stream;
 
 @Rule(key = "S2187")
 public class NoTestInTestClassCheck extends IssuableSubscriptionVisitor {
 
-  private static final Predicate<SymbolMetadata.AnnotationInstance> PREDICATE_ANNOTATION_TEST_OR_UNKNOWN = input -> {
-    Type type = input.symbol().type();
-    return type.isUnknown() || isTestAnnotation(type);
-  };
+  private final Map<String, Boolean> testAnnotations;
```

Comment:
As discussed : while unlikely to happen in this case , this is a memory leak in disguise and goes against the usual way to write checks that are running on a file by file basis.